docs: add section `no context api` (#108)
kitajs/html: README.md

<br />

### Why there is no `context` API?

We choose to not provide a `context` API. Here the reasons why:

This library only outputs strings (or Promises that resolve to strings), we don't manage

> lifecycle or state. So the main selling point to having a `context` API is to avoid prop

Reviewer: ```suggestion
lifecycle or state. So the main selling point to having a `context` API is to avoid [prop drilling](https://www.geeksforgeeks.org/what-is-prop-drilling-and-how-to-avoid-it).
```

Reviewer: unusual term, a reference is good

drilling.

> A context Api would need to use a request identifier (that we are going to call

Reviewer: the closest we can keep the documentation to the actual source code, easier it is to new maintainers help us out.

```suggestion
A context Api would need to use a request identifier (internally called `rid`)
to be safe to access asynchronously. Take for example two requests that are
```

request-id) to be safe to access asynchronously. Take for example two requests that are
being processed at the same time, both of them use some async components. If we use a

> global context without scoping it behind some sort of request-id, the second request would

Reviewer: ```suggestion
global context without scoping it behind some sort of `rid`, the second request would
```

override the context of the first request. So for the first request, some part of the
rendering (the non async part) would use the correct context, but then the other section

> inside the async component would use the context of the second request. This is bad. Each

Reviewer: ```suggestion
inside the async component would use the context of the second request. **This is bad.** Each
```

> component that uses context would need to receive the request-id as a prop. This will

Reviewer: ```suggestion
component that uses context would need to receive the `rid` as a prop. This will
```

> defeat the purpose of having a context in the first place. The request-id is needed to

Reviewer: ```suggestion
defeat the purpose of having a context in the first place. The `rid` is needed to
```

make the context async safe.

The only way to maintain data consistency across concurrent renders without attaching a

> request locator (request-id), is by using

Reviewer: ```suggestion
request locator (`rid`), is by using
```

Our recommendation is to use props. If you want to avoid prop drilling, you can use a
composition style of writing components. This is a common pattern in other JSX-based
libraries, and it works well with this library too.

```tsx
// Layout.tsx

function Doctype(props: Html.PropsWithChildren<{ head: JSX.Element; title?: string }>) {
  return (
    <>
      {'<!doctype html>'}
      <html lang="en">
        <head>
          <meta charset="UTF-8" />
          <meta name="viewport" content="width=device-width, initial-scale=1.0" />
          <title>{props.title || 'Hello World!'}</title>
          {props.head}
        </head>
        <body>{props.children}</body>
      </html>
    </>
  );
}

function SomeLayoutPage(
  props: Html.PropsWithChildren<{ leftSection: JSX.Element; topSection: JSX.Element }>
) {
  return (
    <div>
      <div>{props.topSection}</div>
      <div>{props.leftSection}</div>
      <div>{props.children}</div>
    </div>
  );
}

function SubMenu(props: { request: Request }) {
  // do thing with the request, maybe check if the user is logged in
  const isLoggedIn = !!request.user;
  return <nav>...</nav>;
}

// Example Request handler using the above components

> 

Reviewer: ```suggestion
```

This library only outputs strings (or Promises that resolve to strings), we don't manage
lifecycle or state. So the main selling point to having a `context` API is to avoid
[prop drilling](https://www.geeksforgeeks.org/what-is-prop-drilling-and-how-to-avoid-it).

A context Api would need to use a request identifier (internally called `rid`) to be safe
to access asynchronously. Take for example two requests that are being processed at the
same time, both of them use some async components. If we use a global context without
scoping it behind some sort of `rid`, the second request would override the context of the
first request. So for the first request, some part of the rendering (the non async part)
would use the correct context, but then the other section inside the async component would
use the context of the second request. **This is bad.**. Each component that uses context
would need to receive the `rid` as a prop. This will defeat the purpose of having a
context in the first place. The `rid` is needed to make the context async safe.

The only way to maintain data consistency across concurrent renders without attaching a
request locator (`rid`), is by using

> (ALS)[https://nodejs.org/api/async_context.html#class-asynclocalstorage]. However, this

Reviewer: ```suggestion
[ALS](https://nodejs.org/api/async_context.html#class-asynclocalstorage). However, this
```

approach introduces a lot of overhead and a significant performance penalty.

Our recommendation is to use props. If you want to avoid prop drilling, you can use a
composition style of writing components. This is a common pattern in other JSX-based
libraries, and it works well with this library too.

```tsx

> // Layout.tsx

Reviewer: ```suggestion
```

function Doctype(props: Html.PropsWithChildren<{ head: JSX.Element; title?: string }>) {
  return (
    <>
      {'<!doctype html>'}
      <html lang="en">
        <head>
          <title>{props.title || 'Hello World!'}</title>
        </head>
        <body>{props.children}</body>
      </html>
    </>
  );
}

function SomeLayoutPage(
  props: Html.PropsWithChildren<{ leftSection: JSX.Element; topSection: JSX.Element }>
) {
  return (
    <div>
      <div>{props.topSection}</div>
      <div>{props.leftSection}</div>
      <div>{props.children}</div>
    </div>
  );
}

function SubMenu(props: { user?: User }) {
  // do thing with the user, maybe check if the user is logged in
  return <nav>{props.user ? 'Logged in' : 'Logged out'}</nav>;
}

// Example Request handler using the above components
app.get('/', (request, response) => {

>   response.html(

Reviewer: ```suggestion
  return (
```

same time, both of them use some async components. If we use a global context without
scoping it behind some sort of `rid`, the second request would override the context of the
first request. So for the first request, some part of the rendering (the non async part)
would use the correct context, but then the other section inside the async component would
use the context of the second request. **This is bad.**. Each component that uses context
would need to receive the `rid` as a prop. This will defeat the purpose of having a
context in the first place. The `rid` is needed to make the context async safe.

The only way to maintain data consistency across concurrent renders without attaching a
request locator (`rid`), is by using
(ALS)[https://nodejs.org/api/async_context.html#class-asynclocalstorage]. However, this
approach introduces a lot of overhead and a significant performance penalty.

Our recommendation is to use props. If you want to avoid prop drilling, you can use a
composition style of writing components. This is a common pattern in other JSX-based
libraries, and it works well with this library too.

```tsx
// Layout.tsx

> 

Reviewer: ```suggestion
```

function Doctype(props: Html.PropsWithChildren<{ head: JSX.Element; title?: string }>) {
  return (
    <>
      {'<!doctype html>'}
      <html lang="en">
        <head>

>           <meta charset="UTF-8" />

Reviewer: ```suggestion
```

>           <meta name="viewport" content="width=device-width, initial-scale=1.0" />

Reviewer: ```suggestion
```

          <title>{props.title || 'Hello World!'}</title>

>           {props.head}

Reviewer: ```suggestion
```

        </head>
        <body>{props.children}</body>
      </html>
    </>
  );
}

function SomeLayoutPage(
  props: Html.PropsWithChildren<{ leftSection: JSX.Element; topSection: JSX.Element }>
) {
  return (
    <div>
      <div>{props.topSection}</div>
      <div>{props.leftSection}</div>
      <div>{props.children}</div>
    </div>
  );
}

function SubMenu(props: { user?: User }) {
  // do thing with the user, maybe check if the user is logged in

>   const isLoggedIn = !!props.user;

Reviewer: ```suggestion
```

>   return <nav>...</nav>;

Reviewer: ```suggestion
  return <nav>{props.user ? 'Logged in' : 'Logged out'}</nav>;
```

}

// Example Request handler using the above components
app.get('/', (request, response) => {

>   response.send(

Reviewer: ```suggestion
  response.html(
```

>     <Doctype title="Home" head={<link rel="stylesheet" href="/style.css" />}>

Reviewer: ```suggestion
    <Doctype title="Home">
```

Our recommendation is to use props. If you want to avoid prop drilling, you can use a
composition style of writing components. This is a common pattern in other JSX-based
libraries, and it works well with this library too.

```tsx
// Layout.tsx

function Doctype(props: Html.PropsWithChildren<{ head: JSX.Element; title?: string }>) {
  return (
    <>
      {'<!doctype html>'}
      <html lang="en">
        <head>
          <meta charset="UTF-8" />
          <meta name="viewport" content="width=device-width, initial-scale=1.0" />
          <title>{props.title || 'Hello World!'}</title>
          {props.head}
        </head>
        <body>{props.children}</body>
      </html>
    </>
  );
}

function SomeLayoutPage(
  props: Html.PropsWithChildren<{ leftSection: JSX.Element; topSection: JSX.Element }>
) {
  return (
    <div>
      <div>{props.topSection}</div>
      <div>{props.leftSection}</div>
      <div>{props.children}</div>
    </div>
  );
}

function SubMenu(props: { user?: User }) {
  // do thing with the user, maybe check if the user is logged in
  const isLoggedIn = !!props.user;
  return <nav>...</nav>;
}

// Example Request handler using the above components

> 

Reviewer: ```suggestion
```
